lib/Model/Importer/Binary/ELFImporter.cpp

```diff
@@ -1187,29 +942,125 @@ Error importELF(TupleTree<model::Binary> &Model,
+          return Importer.import();
+        } else {
+          auto MipsImporter = MipsELFImporter<object::ELF32LE,
+                                              true>(Model,
+                                                    TheBinary,
+                                                    PreferredBaseAddress);
+          return MipsImporter.import();
+        }
       } else {
-        return Importer.import<object::ELF32LE, false>();
+        if (!IsMIPS) {
+          auto Importer = ELFImporter<object::ELF32LE,
+                                      false>(Model,
+                                             TheBinary,
+                                             PreferredBaseAddress);
+          return Importer.import();
+        } else {
+          auto MipsImporter = MipsELFImporter<object::ELF32LE,
+                                              false>(Model,
+                                                     TheBinary,
+                                                     PreferredBaseAddress);
+          return MipsImporter.import();
+        }
       }
     } else {
       if (HasRelocationAddend) {
-        return Importer.import<object::ELF32BE, true>();
+        if (!IsMIPS) {
+          auto
+            Importer = ELFImporter<object::ELF32BE, true>(Model,
+                                                          TheBinary,
+                                                          PreferredBaseAddress);
+          return Importer.import();
+        } else {
+          auto MipsImporter = MipsELFImporter<object::ELF32BE,
+                                              true>(Model,
+                                                    TheBinary,
+                                                    PreferredBaseAddress);
+          return MipsImporter.import();
+        }
       } else {
-        return Importer.import<object::ELF32BE, false>();
+        if (!IsMIPS) {
+          auto Importer = ELFImporter<object::ELF32BE,
+                                      false>(Model,
+                                             TheBinary,
+                                             PreferredBaseAddress);
+          return Importer.import();
+        } else {
+          auto MipsImporter = MipsELFImporter<object::ELF32BE,
+                                              false>(Model,
+                                                     TheBinary,
+                                                     PreferredBaseAddress);
+          return MipsImporter.import();
+        }
       }
     }
   } else if (PointerSize == 8) {
     if (IsLittleEndian) {
       if (HasRelocationAddend) {
-        return Importer.import<object::ELF64LE, true>();
+        if (!IsMIPS) {
+          auto
+            Importer = ELFImporter<object::ELF64LE, true>(Model,
+                                                          TheBinary,
+                                                          PreferredBaseAddress);
+          return Importer.import();
```

Comment:
Given that `ELFImporter` is now `virtual`, please make a factory method returning a `std::unique_ptr<ELFImporter>`.

Reply:
Hmm...the ELFImporter is a template, so I am not sure it will improve the code...

Reply:
You can have `ELFImporterBase` that's non-templated. Also, you don't have to write `Importer.import()` several times. If we have add methods it's going to be helpful.
Overall, if you have a virtual class, having a factory method returning a `unique_ptr` is a pretty common practice.